### DISCUSSION TITLE
Add file behind the scenes to vscode workspace

### DISCUSSION THREAD

**Discussion by avii-ahuja** [TypeFox/monaco-languageclient#579] [Q&A]
I've been following the React example and I want to add an XML file behind the scenes to the workspace. This is because the language server that I'm using needs to have an XML catalog file in the workspace to work correctly. How do i do this? So far, I've done something like this - 
[https://github.com/avii-ahuja/monaco-languageclient/blob/main/packages/examples/src/common/client-commons.ts](https://github.com/avii-ahuja/monaco-languageclient/blob/main/packages/examples/src/common/client-commons.ts)
```typescript
// create the model
    const uri = Uri.parse("/workspace/vocabulary/pizza.oml");
    const modelRef = await createModelReference(uri, config.content);
    modelRef.object.setLanguageId("oml");

    //     const catalogRef = await createModelReference(
    //         Uri.parse("/workspace/catalog.xml"),
    // "
    //     );
    // const catalogCode = fs.readFileSync(
    //     "/Users/avii/Desktop/monaco-languageclient/packages/examples/src/common/catalog.xml",
    //     "utf-8"
    // );
    const rawText = await fetch(raw);
    const xmlText = await rawText.text();
    const catalogRef = await createModelReference(
        Uri.parse("/workspace/catalog.xml"),
        xmlText
    );
```
pizza.oml is the file being edited, and I need to insert catalog.xml in the workspace (but it should be invisible to the editor).

**Comment by cdietrich** [comment]
> This is because the language server that I'm using needs to have an XML catalog file in the workspace to work correctly

when and how does it need it

**Reply by avii-ahuja** [reply]
The language server is for the OML (Ontological Modeling Language) [https://github.com/opencaesar/oml](https://github.com/opencaesar/oml)
The language resolves URLs to local file paths via a `catalog.xml` file which needs to be in the workspace. 
Sample catalog file code - 
```xml
<?xml version="1.0"?>
<catalog xmlns="urn:oasis:names:tc:entity:xmlns:xml:catalog" prefer="public">
	<rewriteURI uriStartString="http://example.com/tutorial1/" rewritePrefix="src/oml/example.com/tutorial1/"/>
	<rewriteURI uriStartString="http://" rewritePrefix="build/oml/"/>
</catalog>
```

This resolves the `http://example.com/tutorial1/vocabulary/pizza#` to a local path such as `src/oml/example.com/tutorial1/vocabulary/pizza.oml` inside this `.oml` file
located at 
 - [https://github.com/opencaesar/pizza-example/blob/master/src/oml/example.com/tutorial1/vocabulary/pizza.oml](https://github.com/opencaesar/pizza-example/blob/master/src/oml/example.com/tutorial1/vocabulary/pizza.oml)
```
@dc:description "A vocabulary about pizzas"
vocabulary <http://example.com/tutorial1/vocabulary/pizza#> as pizza {
	
	extends <http://www.w3.org/2001/XMLSchema#> as xsd
```
(rest of file excluded)

More information here where I got the OML pizza example project from -[https://github.com/opencaesar/pizza-example/tree/master](https://github.com/opencaesar/pizza-example/tree/master)

**Reply by cdietrich** [reply]
But it does it on the fly and not on initialize call on the workspace?

**Reply by cdietrich** [reply]
And shouldn’t you open the pizza file as well ?

**Reply by avii-ahuja** [reply]
Ideally, i need to open the pizza file as well. I'm not familiar with how to do that and appreciate any advice on how to proceed/what to try out.

**Reply by cdietrich** [reply]
The same way you open catalog file ?

btw did you debug the server ?

**Reply by avii-ahuja** [reply]
Yes, I'm sure the server is bug-free. Do I need to open the catalog file for it to get "acknowledged"/"seen" by the language server? Can't it just be in the workspace and the server would see it?

**Reply by cdietrich** [reply]
Btw the code you shared is Xtext and there a separate Java process is needed . How do you start the server ?

**Reply by avii-ahuja** [reply]
I start the server using by running a jar file. I can confirm the server is working as it sends back annotations and code hints. I use an express server to accept websockets that also starts the language server. Every notification that is received by the express server is forwarded to the language server.

**Reply by cdietrich** [reply]
Then the question is when the server reads the xml

I propose you start it with remote debugging enabled and debug if the did open events are seen and how they are treated

**Reply by avii-ahuja** [reply]
Is the method i'm using enough to put the file in the workspace? And if I need to unzip and put a folder in the workspace, is there an easier way to do it?

I'll look into remote debugging

**Reply by cdietrich** [reply]
And also what server does on initialize

**Reply by cdietrich** [reply]
No it does not put anything in workspace

**Reply by cdietrich** [reply]
Not physically and not early

**Reply by avii-ahuja** [reply]
How do i put the catalog file in the workspace then? if i don't want to show it in the editor.

**Reply by cdietrich** [reply]
This is semantics of client and server
Which workspace folder does the server use

**Reply by cdietrich** [reply]
And does it support live editing of the xml

**Reply by avii-ahuja** [reply]
It doesn't support editing the XML. It only supports OML. As for the workspace, it looks at the OML file being edited - in this case - it looks at `pizza.oml`. So, it would search the directories starting from the parent of `pizza.oml` for a `catalog.xml` file. I hope this makes it clearer?

**Reply by avii-ahuja** [reply]
so the folder of`pizza.oml` and its parent folders

**Reply by cdietrich** [reply]
=> the xml must be present on the initialize call then ?!?
and what does it for the oml file you edit? which directory does it map the workspace to?

**Reply by avii-ahuja** [reply]
Sorry, I'm not familiar with the initialize call. Can you tell me where to find it?

I'm using the file:///workspace/<files>
as my workspace

**Reply by cdietrich** [reply]
the problem is that the catalog thing seems to be highly custom to the impl of the server.
i dont think they do it on initialize but on the fly.

=> ootb this wont work unless you put the catalog.xml to what ever OmlUriResolver
does think about the uri of the oml file you open.
if that one is 

" file:///workspace/"
you need to put it there

**Reply by cdietrich** [reply]
all the calls go though LanguageServerImpl on server side

**Reply by avii-ahuja** [reply]
I'm trying to put the file there, my code has the xml catalog file in the workspace. Is there another way to put the catalog file in the workspace or am I misunderstanding?

**Reply by cdietrich** [reply]
again. this is convention when you start the server.
why dont you put the file in the workspace when you start the server?

the idea of the protocol is that server and client share the workspace and everything else is convention between client and server

**Reply by cdietrich** [reply]
maybe you can use a command or a custom lsp extension (server and client side) to pass the config to the server and write it there
